### PR TITLE
Dl 590

### PR DESCRIPTION
**Pull request by robert-webb-hmrc**

### CODE REVIEW

File: app/views/govuk_wrapper.scala.html [hunk 1]
```diff
@@ -99,3 +99,12 @@
 }
 
 @hmrcGovUkTemplate(Some(title), bodyClasses)(head, bodyEnd, insideHeader, afterHeader, footerTop, Some(uiLayouts.footer_links()), true)(content)
+
+@isWelsh = @{
```
Comment (Ufuoma-Daniel) [inline app/views/govuk_wrapper.scala.html]: what is this for? This isn't referenced anywhere.

File: app/views/helpers/questionAnswerRow.scala.html [hunk 2]
```diff
@@ -31,7 +32,7 @@
                     case string: String => string
                     case int: Int => int.toString
                     case bigDecimal: BigDecimal => s"£${MoneyPounds(bigDecimal, decimalPlaces).quantity}"
-                    case date: LocalDate => datePageFormatNoZero.format(date)
+                    case date: LocalDate => date.localFormat("d MMMM uuuu")
```
Comment (Ufuoma-Daniel) [inline app/views/helpers/questionAnswerRow.scala.html]: d MMMM uuuu shouldBe d MMMM yyyy?

File: app/views/helpers/questionAnswerRowNoLink.scala.html [hunk 3]
```diff
@@ -29,7 +30,7 @@
         case string: String => string
         case int: Int => int.toString
         case bigDecimal: BigDecimal => s"£${MoneyPounds(bigDecimal, decimalPlaces).quantity}"
-        case date: LocalDate => datePageFormatNoZero.format(date)
+        case date: LocalDate => date.localFormat("d MMMM uuuu")
```
Comment (Ufuoma-Daniel) [inline app/views/helpers/questionAnswerRowNoLink.scala.html]: uuuu -> yyyy?

File: test/assets/MessageLookup.scala [hunk 4]
```diff
@@ -22,6 +22,7 @@ import common.Constants
 object MessageLookup {
 
   // TO MOVE
+  //TODO: need to investigate why commented out vals are not being used in ANY tests
```
Comment (Ufuoma-Daniel) [inline test/assets/MessageLookup.scala]: There is a todo here that needs to be completed? Lots of commented out code in this file.

File: test/constructors/DeductionDetailsConstructorSpec.scala [hunk 5]
```diff
@@ -326,9 +328,10 @@ class DeductionDetailsConstructorSpec extends UnitSpec with WithFakeApplication
         lazy val result = DeductionDetailsConstructor.privateResidenceReliefDaysClaimedBeforeRow(
           Some(PrivateResidenceReliefModel("Yes", Some(4), None)), edgeDatesJustDaysBefore)
 
-        lazy val beforeString = Dates.dateMinusMonths(edgeDatesJustDaysBefore.disposalDateModel, 18)
+        lazy val beforeDate = Dates.dateMinusMonths(edgeDatesJustDaysBefore.disposalDateModel, 18)
```
Comment (Ufuoma-Daniel) [inline test/constructors/DeductionDetailsConstructorSpec.scala]: unused value.

File: test/util/MessagesSpec.scala [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+        welshMessages.keySet shouldBe englishMessages.keySet
+      }
+    }
+
+    "have a non-empty message for each key" in {
+      assertNonEmpty("English", englishMessages)
+      assertNonEmpty("Welsh", welshMessages)
+    }
+
+    "have no unescaped single quotes in value" in {
```
Comment (Ufuoma-Daniel) [inline test/util/MessagesSpec.scala]: this test isn't working?

File: app/common/Constants.scala [hunk 7]
```diff
@@ -19,4 +19,5 @@ package common
 object Constants {
   val maxNumeric: BigDecimal = 1000000000
   val maxLettingsRelief: BigDecimal = 40000
+  val maxAllowance: BigDecimal = 11100
```
[the remark this answers is not in the document]
Comment (Tom-Stacey) [reply app/common/Constants.scala]: My question would be more 'why isn't this being used anymore?' :)

File: app/common/Validation.scala [hunk 8]
```diff
@@ -52,6 +54,29 @@ object Validation {
...
+                                  errMsgKey: String = "calc.common.error.maxNumericExceeded"
+                                ): Constraint[BigDecimal] = Constraint("constraints.maxValue")({
+    value => maxMoneyCheck(value, maxValue, errMsgKey)
+  })
+
+  def maxMonetaryValueConstraint[T](maxValue: BigDecimal, extractMoney: T => Option[BigDecimal]): Constraint[T] = {
+    Constraint("constraints.maxValueCustom")({
+      data => extractMoney(data).map {
+        maxMoneyCheck(_, maxValue, "calc.common.error.maxNumericExceeded")
+      }.getOrElse(Valid)
```
[the remark this answers is not in the document]
Comment (Tom-Stacey) [reply app/common/Validation.scala]: I'm guessing that getOrElse(valid) is ok as if it's an optional field we don't want it to be returned as invalid when not completed

File: app/common/Constants.scala (hunk 7 above)
[the remark this answers is not in the document]
Comment (robert-webb-hmrc) [reply app/common/Constants.scala]: Added for MaxAEA error message check, reverted once maxAEA was implemented instead.

[the remark this answers is not in the document]
Comment (Ufuoma-Daniel) [reply app/common/Constants.scala]: Yeah the max allowance is a value that comes from a call to the backend so it was a misunderstanding to create a constant for maxAllowance here anyways.

File: app/views/calculation/acquisitionDate.scala.html [hunk 9]
```diff
@@ -17,22 +17,17 @@
...
 @main_template(Messages("calc.acquisitionDate.question")) {
 
     <a id="back-link" class="back-link" href="@controllers.routes.DisposalCostsController.disposalCosts.url">@Messages("calc.base.back")</a>
 
-    @defining(
-        if(acquisitionDateForm.hasErrors) {
-            if(acquisitionDateForm.errors.exists(_.message.contains(Messages("calc.common.date.error.invalidDate"))))
-                Some(Messages("calc.acquisitionDate.errorIncompleteDate"))
-            else Some(Messages("calc.acquisitionDate.errorFutureDate"))
-        } else None
```
[the remark this answers is not in the document]
Comment (robert-webb-hmrc) [reply app/views/calculation/acquisitionDate.scala.html]: Not sure what you mean here.

File: test/util/MessagesSpec.scala (hunk 6 above)
Comment (robert-webb-hmrc) [reply test/util/MessagesSpec.scala]: Test still functioning on my machine. The quote is looked for within the message values, so any quotes entered outside the values do not count.

File: test/assets/MessageLookup.scala (hunk 4 above)
Comment (robert-webb-hmrc) [reply test/assets/MessageLookup.scala]: Not part of this story.

File: app/constructors/PurchaseDetailsConstructor.scala [hunk 10]
```diff
@@ -63,24 +63,24 @@ object PurchaseDetailsConstructor {
     Some(QuestionAnswerModel(
```
[the remark this answers is not in the document]
Comment (Tom-Stacey) [reply app/constructors/PurchaseDetailsConstructor.scala]: It's a good spot, but I think it was agreed that we would leave these as a separate story - the MVP for Welsh was to get it working in Welsh, I think I remember, so we said we'd handle styling issues later. We should probably make sure those stories have been raised before this gets moved to Done, though

File: test/constructors/DeductionDetailsConstructorSpec.scala (hunk 5 above)
Comment (robert-webb-hmrc) [reply test/constructors/DeductionDetailsConstructorSpec.scala]: deleted

File: app/views/helpers/questionAnswerRowNoLink.scala.html (hunk 3 above)
Comment (robert-webb-hmrc) [reply app/views/helpers/questionAnswerRowNoLink.scala.html]: changed

File: app/views/helpers/questionAnswerRow.scala.html (hunk 2 above)
Comment (robert-webb-hmrc) [reply app/views/helpers/questionAnswerRow.scala.html]: changed

File: app/forms/AnnualExemptAmountForm.scala [hunk 11]
```diff
@@ -16,30 +16,28 @@
 
 package forms
 
+import common.Constants
```
[the remark this answers is not in the document]
Comment (robert-webb-hmrc) [reply app/forms/AnnualExemptAmountForm.scala]: deleted

File: app/forms/AcquisitionMarketValueForm.scala [hunk 12]
```diff
@@ -22,23 +22,20 @@ import common.Validation._
 import models.AcquisitionValueModel
 import play.api.data.Forms._
 import play.api.data.Form
-import play.api.i18n.Messages
+import play.api.i18n.Lang
```
[the remark this answers is not in the document]
Comment (robert-webb-hmrc) [reply app/forms/AcquisitionMarketValueForm.scala]: removed

File: app/views/govuk_wrapper.scala.html (hunk 1 above)
Comment (robert-webb-hmrc) [reply app/views/govuk_wrapper.scala.html]: removed

File: app/views/calculation/acquisitionDate.scala.html (hunk 9 above)
[the remark this answers is not in the document]
Comment (Ufuoma-Daniel) [reply app/views/calculation/acquisitionDate.scala.html]: The error values messages "calc.acquisitionDate.errorIncompleteDate" and "calc.acquisitionDate.errorFutureDate" are not being used anymore and were replaced with an invalidDate key.

File: test/util/MessagesSpec.scala (hunk 6 above)
Comment (Ufuoma-Daniel) [reply test/util/MessagesSpec.scala]: I've retested it and it works if the single quote is in the middle of the message value. If it is at the start or end of the message then it doesn't notice it.

File: test/assets/MessageLookup.scala (hunk 4 above)
Comment (robert-webb-hmrc) [reply test/assets/MessageLookup.scala]: Done

File: app/views/calculation/acquisitionDate.scala.html (hunk 9 above)
[the remark this answers is not in the document]
Comment (robert-webb-hmrc) [reply app/views/calculation/acquisitionDate.scala.html]: removed

**Pull request merged**